src/snowflake/connector/cache.py

```diff
@@ -260,3 +319,218 @@ def _expiration(self, k: K) -> None:
+
+
+class SFDictFileCache(SFDictCache):
+    def __init__(
+        self,
+        file_path: str | dict[str, str],
+        entry_lifetime: int = constants.DAY_IN_SECONDS,
+        file_timeout: int = 0,
+    ) -> None:
+        """Inits an SFDictFileCache with path, lifetime.
+
+        File path can be a dictionary that contains different paths for different OSes,
+        possible keys are: 'darwin', 'linux' and 'windows'. If a current platform
+        cannot be determined, or is not in the dictionary we'll use the first value.
+
+        Once we select a location based on file path, we write and read a random
+        temporary file to check for read/write permissions. If this fails OSError might
+        be thrown.
+        """
+        super().__init__(
+            entry_lifetime=entry_lifetime,
+        )
+        if isinstance(file_path, str):
+            self.file_path = os.path.expanduser(file_path)
+        else:
+            current_platform = platform.system().lower()
+            if current_platform is None or current_platform not in file_path:
+                self.file_path = next(iter(file_path.values()))
+            else:
+                self.file_path = os.path.expanduser(file_path[current_platform])
+        # Once we decided on where to put the file cache make sure that this
+        #  place is readable/writable by us
+        random_string = "".join(random.choice(string.ascii_letters) for _ in range(5))
+        tmp_file, tmp_file_path = tempfile.mkstemp(
+            dir=os.path.dirname(self.file_path),
+        )
+        try:
+            with open(tmp_file, "w") as w_file:
+                # Might raise an OSError
+                w_file.write(random_string)
+            with open(tmp_file_path) as r_file:
+                # Might raise an OSError
+                if r_file.read() != random_string:
+                    Exception("Temporary file just written has wrong content")
+        finally:
+            if os.path.exists(tmp_file_path) and os.path.isfile(tmp_file_path):
+                os.unlink(tmp_file_path)
+        self.file_timeout = file_timeout
+        self._file_lock_path = f"{self.file_path}.lock"
+        self._file_lock = FileLock(self._file_lock_path, timeout=self.file_timeout)
+        self.last_loaded: datetime.datetime | None = None
+        if os.path.exists(self.file_path):
+            self._load()
+
+    def __getitem__(self, k: K) -> V:
+        """Returns an element if it hasn't expired yet in a thread-safe way."""
+        self._lock.acquire()
+        # TODO: This variable could be replaced by a wrapper class that keeps track
+        #  of whether the lock is locked, but unless this function gets extended I
+        #  feel like it's an overkill. Make sure to change the bool right after
+        #  self._lock.acquire() and self._lock.release().
+        currently_holding = True
+        try:
+            try:
+                t, v = self._cache[k]
+            except KeyError:
```

Comment:
Curious: did we choose `try-except` instead of key lookup first in `if` because of performance reasons? Just feel a bit awkward with the nested `try-except`.